### ISSUE TITLE
Guard `plt.subplot` kwargs a bit better?

### ISSUE THREAD

**Issue by jklymak** [matplotlib/matplotlib#15964]
### Bug report

A user on [stackoverflow](https://stackoverflow.com/questions/59377017/matplotlib-simplest-possible-example-for-2-x-1-grid-of-plots-fails-with-cann) did:

```python
from matplotlib import pyplot as plt
fig, (ax1, ax2) = plt.subplot(nrows=1, ncols=2, sharey=True)
```

which of course is trying to use `subplots` but instead calls `subplot`.  The error message is quite mystifying, however:

```
...
   ~/.virtualenvs/tcpsp/lib/python3.7/site-packages/matplotlib/cbook/__init__.py in join(self, a, *args)
        923 
        924         for arg in args:
    --> 925             set_b = mapping.get(weakref.ref(arg), [weakref.ref(arg)])
        926             if set_b is not set_a:
        927                 if len(set_b) > len(set_a):


    TypeError: cannot create weak reference to 'bool' object
```

While I'm not suggesting we should catch every possible user misuse of kwargs, it seems this basic entry point should have a few more guardrails.

**Comment by KartikChugh** [comment]
` plt.subplot(3, 3, i+1, sharey=True)` still throws the error for me

**Comment by jklymak** [comment]
@KartikChugh Can you be more specific, including version of matplotlib? Note the fix only went in 3.3.0

**Comment by KartikChugh** [comment]
Oh I see, assumed Google Colab had the latest version. Mine is on 3.2.2.
